### ISSUE TITLE
Property does not exist on type 'never' when it exists

### ISSUE THREAD

**Issue by wgebczyk** [microsoft/TypeScript#38286]
Similar issue was reported #27041, but last suggestion is to go to StackOverflow to look for help -- usually you can get there hacks and partial solutions. This case seems to be legit and I would like to understand why TS works as designed and I'm wrong.

In short, in case of guard function for "empty" interface (extending another non empty one) starts narrow type to never.

To demonstrate that there might be code that works and does something non trivial (runtime verification of interface implementation). I've tried to keep sample as small as possible to demonstrate issue and prove that there might be significant reason for writing such code.

I do know workaround -- just move if over guard for "empty" interface as last one (what to do when there will be more that one empty interface in inheritance chain)?

**Code**

```ts
interface IModifier {
  readonly id: string;
}
interface IXModifier extends IModifier {
    name: string;
}
interface IYModifier extends IModifier {
}


const symbolModifier = Symbol("IModifier");
function isModifier(o: any): o is IModifier {
    return Object.is(o[symbolModifier], symbolModifier);
}
function implementsModifier(o: IModifier) {
    (o as any)[symbolModifier] = symbolModifier;
}


const symbolXModifier = Symbol("IXModifier");
function isXModifier(o: any): o is IXModifier {
    return Object.is(o[symbolXModifier], symbolXModifier);
}
function implementsXModifier(o: IXModifier) {
    (o as any)[symbolXModifier] = symbolXModifier;
}


const symbolYModifier = Symbol("IYModifier");
function isYModifier(o: any): o is IYModifier {
    return Object.is(o[symbolYModifier], symbolYModifier);
}
function implementsYModifier(o: IYModifier) {
    (o as any)[symbolYModifier] = symbolYModifier;
}


abstract class Modifier implements IModifier {
    readonly id: string;
    constructor(id: string) {
        implementsModifier(this);
        this.id = id;
    }
}
class XModifier extends Modifier implements IXModifier {
    readonly name: string;
    constructor(id: string, name: string) {
        super(id);
        implementsXModifier(this);
        this.name = name;
    }
}
class YModifier extends Modifier implements IYModifier {
    constructor(id: string) {
        super(id);
        implementsYModifier(this);
    }
}


function tsIsBrokenBeyondAllRepair(arr: IModifier[]): void {
    for (const m of arr) {
        if (isYModifier(m)) {
            console.log(`[Y] /${m.id}/`);
        } else if (isXModifier(m)) {
            console.log(`[X] /${m.id}/${m.name}/`);
        } else {
            throw new Error("Not supported.");
        }
    }
}

const modifiers = [
    new XModifier("X.1", "X--1"),
    new XModifier("X.2", "X--2"),
    new YModifier("Y.1"),
    new YModifier("Y.2"),
];
tsIsBrokenBeyondAllRepair(modifiers);
```

**Expected behavior:**

Well, if guard says that some value is of specific type then TS should stop pretend that value is of more narrow type to do less work. 

**Actual behavior:**

In bigger sequence of if-else TS starts to think that type is "never" and happily cuts the tedious verification work and gives up

**Playground Link:**

[Playground Link](https://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=80&pc=1#code/JYOwLgpgTgZghgYwgAgJIFkD2ATYNjTIDeAUMslBHNpiADYCeyw2AXMgM5hSgDmA3CQC+JUJFiIUqABpZc+QhAAekENg5o5eAlGJlyyEHAC2Edlx4gBw0eGjwkaAJpaFu5avWac2wqREkJAi0XJwMxgBGmHSuOsgAvMgAyuFRdAAUAEQYPm6ZAJSCMACuIAhgwLTMHLHQ6ZjscCAM+eyY1d7ycaQGFBBgxVAgyADyEQBWEOUAdMAc9QDaHKnRtVAAugA0YZGruTqFNiVlFVXAxgAOdBCm4DX7dQ2dvlD5er31yHAaTS1LKzEHhsEjs0mtBAEgiEwKDorIgSCUrsMtl4V1oAUiqVypVhnM0S96o1mq1kO05mgCW53gZKAMhqMJlMwLN5ph-siqTotrC6FzoIcRMccWdLtdbmAOPyoETKWs3j0DJ9vl8SRy0tL1iDlpzwTZAsEQKEdWkXAjEki0llUGb0VBMSRhac8RxbYSnr9SeSNDa1jTyHTBsMxpMZnNFibom63DzI3RowcIY7sc7mGKbhA7gnHuxfUCFfpyMqfmq49ngYky3rIXAIhZEDCEHRvho-ecrhm7s9qYqA1QaPQmCxzNw+IJeoaLMVypgZcPOKOrAXer12+LM5K1ukwAALOaHFcGXdzWbYEEsccGEQiJst5DS5AeTNeNvpiU+h+9vrUWiMQwmMwF0sawJ2hKBpzAWd0nnCw+G2IxTBHYDl0PThiguOoWAPVC107SVpW3PcOGww9jw4aYEJQRJKMvchryCZsOA0ctHxUZ9WwRXD32cP0v0nbgIKgmDF14FDDw4dDMOwEiVy4jdXS3MiZPowInVxZBJVQDgACEoEwABrTNtIgBhaGwABBOg6AAJQgC44GAGU4CgKBczWBZ1lJAA3TAWH9ZAYFnZB0n45BjDJGAvhcsTZMi6CFKBdJjHyGLUOQSdoggaY6EwXh0gAAwWJwtQAegAEiIYxTyEEr8pk3ohEfOgOBQPBgvxLdktS1CMuubLcoKhZpFKiqqpYGrRoogCarq2iV0aiBmpQL80t3PSAHdDAgTaAFEXKgzIADlMBhCSLguWdIGwaYHTSkQGv1ULjCBDREgWQsts2gjMmkaYAEZMm2H6AFpgYB-JNg+kBtvvLcfumAAmQHkBB4GkYhqGYfLLInH+gpId6aHNuxzJcfRyH1kETSdL0wyQGM0y1Esmy7IcmVnrtYjBCAA)

**Related Issues:** 

Property does not exist on type 'never' reported when it shouldn't #27041

**Comment by nmain** [comment]
Your interface types are all structural, so:
1. The `IModifier[]` annotation on `arr` means that `m` conforms to `{ readonly id: string }`
2. The if branch of `isYModifier(m)` means that `m` conforms to `{ readonly id: string }`
3. The else branch of `isYModifier(m)` means that `m` does not conform to `{ readonly id: string }`
4. Combining 1 and 3, `m` must be `never` in the else branch since it's impossible to create a value which both has and doesn't have that `id` property.

**Comment by RyanCavanaugh** [comment]
@nmain is correct

User-defined type predicates should not return values that don't conform to the structural type meanings of those types; doing so will lead to contradictions as you've demonstrated.

**Comment by wgebczyk** [comment]
1. Then how to model interface base modelling? Tried to use mixins but failed as is enforce ctors with "...args:any[]". Then I've switched whole context to private symbols for each interface, but this limitation of interfaces (you call them structural, hence expected by me would be nominal).

2. Well... could you elaborate a bit more on "User-defined type predicates"? My docs reading is:
https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

"TypeScript will narrow that variable to that specific type if the original type is compatible.".
Looking at signature of function "function isXModifier(o: any): o is IXModifier", I take type "any" and explicitly narrow that to "IXModifier".

```
1. if (isYModifier(m)) {
2.   console.log(`[Y] /${m.id}/`);
3. } else if (isXModifier(m)) {
4.   console.log(`[X] /${m.id}/${m.name}/`);
5. } else {
6.   throw new Error("Not supported.");
7. }
```
In line 3 I do express that type is "IXModifier", hence line 4 should see it there. I agree that in line 3 "between else and if" it might be never or whatever.

**Comment by InExtremaRes** [comment]
Coming to TypeScript from other languages can be a little confusing, I pass through this as well.

I think you're falling into two different "traps" here.

First, type predicates. A type predicate narrows some type, but you have to consider what others types are at play before, as @RyanCavanaugh said, to avoid contradictions. Consider this example:

```ts
declare function isString(value: any): value is string;

function f1(value: number) {
	if (isString(value)) {
    	value; // <-- never
	}
}
```

Inside `f1` you have first stated that `value` _is_ of type `number`, the type predicate tries to _narrow_ that into `string` but then it should be of type `number & string`, a type that cannot exist, so the resulting type is `never` or otherwise the laws of the universe are broken and anything is possible...

If you know beforehand that the type can really by _anything_ you can declare that explicitly, saying that `value` is of type `any`, `unknown`, `number | string` or something like that:

```ts
function f2(value: unknown) {
	if (isString(value)) {
		value; // <-- string :)
	}
}
```

Something similar happens with another type predicate that is not user-defined:

```ts
function f3(value: number[]) {
	if (value instanceof Map) {
		value; // <-- number[] & Map<any, any>
	}
}
```

In the last example the inferred type is `number[] & Map<any, any>` because... well, that is what you have said to the type system until now. In this case the type is not `never` because it is possible to construct such a type. So you can narrow a type with a type predicate but no without taking into account the current type of the value. Of course everything I said here doesn't _have to be_ this way, is a design choice, but I think is the right one; you can always start from a wider type or make an assertion if you are really sure.

The second one is the structural conformance. TypeScript types are not nominal (there are some exceptions, like [this](https://github.com/microsoft/TypeScript/wiki/FAQ#when-and-why-are-classes-nominal)) so you cannot make interface base modeling the way you can with other languages, at least not in the same way.

In your case you have:

```ts
interface IYModifier extends IModifier {
}
```

In other words, to the eyes of TypeScript, `IYModifier` is basically the same as `IModifier` (in the sense that any one is assignable to the other), and therefore when you try to narrow to `IYModifier` it is the same as narrowing to `IModifier`, that is what @nmain has correctly said before. In your code:

```ts
function tsIsBrokenBeyondAllRepair(arr: IModifier[]): void {
    for (const m of arr) {
        if (isYModifier(m)) { // <-------------- (A)
            console.log(`[Y] /${m.id}/`);
        } else if (isXModifier(m)) { // <------- (B)
            console.log(`[X] /${m.id}/${m.name}/`);
        } else {
            throw new Error("Not supported.");
        }
    }
}
```

In (A), inside the _then_ branch, you are stating that `m` is of type `IYModifier`, but since that is structurally the same as `IModifier`, is like you were narrowing to `IModifier` instead. Then in (B) you are in the _else_ branch so `m` have not pass the first predicate and cannot be of type `IModifier`. Thus, `m` is of type `IModifier` (by the type of the parameter `arr`) and is not of type `IModifier` (by the type predicate), which is a contradiction and the resulting type is `never`.

The problem, as said before, is that `IYModifier` is not different from `IModifier` so probably it shouldn't exist. If `IYModifier` has some other property that makes it different from `IModifier` then the code above will work since there is no contradiction anymore.

So, well, I have said nothing different from what has been said in the comments before, but I hope that can help you. TypeScript type system _is structural_. The [FAQ](https://github.com/microsoft/TypeScript/wiki/FAQ) have a lot more info and rationales behind this. Of course you can disagree with this design choice, but it is at core of TypeScript and it is what it is.

There are some tricks to make a type nominal if you really need to express that, but maybe you should redesign your interfaces instead of fight the type system you are using (but I don't know your use cases). Why is `IYModifier` a thing if it have the same properties as `IModifier`?

Really hope this helps. I'm not trying to make a debate here, just sharing my experience if it worth something. If my understanding and explanation is wrong I appreciate being corrected.

Regards.

**Comment by wgebczyk** [comment]
Thanks for explanations! You are right, either you bend to language that is using you, or you try to choose lang that fits your needs.
